DataLinks: Sanitize data/panel link URLs

dprokop opened the pull request:
Fixes #20974

Changes introduced:
1. Both panel and data link urls are now sanitized using [@braintree/sanitize-url](https://github.com/braintree/sanitize-url). ~~Unfortunately the xss package we use can't do it on a string that is not a html~~. xss package that we use exposes `safeAttrValue` which could be used, but it also performs html escaping, which we want to delegate to components that actually render the html.
2. Added `sanitize-url.d.ts` file for the package type safety
3. Introduced possibility to mock app config in the tests

dprokop committed:
Sanitize html in panel links

dprokop committed:
Add sanitize-url package

dprokop committed:
Enable config mocking

dprokop committed:
Sanitize datalinks urls

dprokop commented on public/app/core/config.ts:
```diff
@@ -1,5 +1,18 @@
+};
+
+export const mockConfig = (mock: Partial<GrafanaBootConfig>) => {
+  configMock = mock as GrafanaBootConfig;
```
```suggestion
  configMock = mock;
```

dprokop committed:
Update public/app/core/config.ts

dprokop commented on public/app/features/panel/panel_ctrl.ts:
```diff
@@ -297,7 +297,7 @@ export class PanelCtrl {
     html += '</div>';
 
-    return html;
+    return config.disableSanitizeHtml ? html : sanitize(html);
```
This sanitizes Angluar panel's links

torkelo commented on public/app/core/config.ts:
```diff
@@ -1,5 +1,18 @@
+  configMock = mock;
+  return () => {
+    configMock = null;
+  };
```
do we need the configMock object? Can't we have a setConfig function instead that set's the config variable?

dprokop replied:
@torkelo mind the type. This allows partial config mocking, only what you need for your test. That's why I opted for "mock" naming

torkelo replied:
can't it be just
```
export function updateConfig(updates: Partial<GrafanaBootConfig>) {
   Object.assign(config, mock);
}
```

dprokop replied:
I like this!

torkelo replied:
This function: 

```ts
export const getConfig = () => {
  return configMock || config;
};
```

feels wrong to have in prod code.

dprokop replied:
@torkelo it's no longer there

torkelo commented on public/app/features/panel/panellinks/specs/link_srv.test.ts:
```diff
@@ -137,4 +138,38 @@ describe('linkSrv', () => {
+
+        // console.log(link);
+        expect(link).toBe(expected);
+        restoreConfig();
```
are not all jest test run in some form of isolation? do we need to the restoreConfig here?

dprokop replied:
Good point. In this case it is not necessary. But if other parts of `link_srv` were dependant on `config` and some other property than `disableSanitizeHtml` than not restoring the partial config might have caused the tests to fail. So I consider this more of a good practice rather than necessity. WDYT?

torkelo replied:
having to restore state after asserts looks to me like a smell / bad test design, I would add a common setup that set's the expected state. or have a common setup / scenario function that let you pass in expected config.

torkelo reviewed (approved):
Looks good, the config mocking & reset feels a bit off , wonder if there is a simper way

hugohaggmark reviewed (approved):
LGTM!

dprokop committed:
Minor test update

dprokop committed:
Merge branch 'fix/20974' of github.com:grafana/grafana into fix/20974

DanCech commented:
The function used by the sanitize library is https://github.com/leizongmin/js-xss/blob/master/lib/default.js#L147 which would avoid the additional import.

dprokop committed:
Remove sanitize-url dependency

dprokop committed:
Remove typings

dprokop committed:
Review update

dprokop commented:
@DanCech Thank you for this suggestion, I have removed the new dependency and used the xss one. 

@torkelo also updated with your suggestion about `updateConfig`

DanCech commented on public/app/features/panel/panellinks/link_srv.ts:
```diff
@@ -209,7 +211,7 @@ export class LinkSrv implements LinkService {
       },
     });
-
+    info.href = getConfig().disableSanitizeHtml ? info.href : sanitizeUrl(info.href);
```
The only problem I have with this approach is that `info.href` will not only be sanitized against XSS-type urls, but it will also be html-escaped.  The problem with that is that when we now put that value into html, we have to be aware that it is already escaped, and not escape it again (It will contain `&lt;` in place of `<` and if we escape that again we'd end up with `&amp;lt;`).

dprokop replied:
I think we can make this assumption as consumers of this data links model should not care about any escaping IMO. One problematic place could be the panel links corner where we actually do the escaping. I'm gonna take a look at that. 

But I'd rather have both sanitisation and escaping here  than scattering this logic across all of the components that use this model.

DanCech replied:
I agree, the best solution would be for the service to handle sanitization so it gives a "trusted" url, but let the html-escaping happen when actually building the html so that it's clear what's going on and so that we don't have issues eg with react escaping the values when embedded in jsx

dprokop committed:
Revert "Remove sanitize-url dependency"

This reverts commit c4f38e6de64b36fb398f0dc61906ae070d69c5ed.

dprokop committed:
Revert "Remove typings"

This reverts commit 676d47e8c2e46d6eeb57003d0a7e0fd99e75310f.

dprokop committed:
Sanitaze, don't escape html when sanitizing URL

torkelo reviewed (approved):
Looks good!
